src/Microsoft.AspNetCore.Mvc.Core/Formatters/OutputFormatter.cs

```diff
@@ -53,19 +53,36 @@ protected virtual bool CanWriteType(Type type)
             if (contentType == null)
             {
                 // If contentType is null, then any type we support is valid.
-                return SupportedMediaTypes;
+                contentType = "*/*";
             }
-            else
-            {
-                List<string> mediaTypes = null;
+            
+            List<string> mediaTypes = null;
+
+            var parsedContentType = new MediaType(contentType);
 
-                var parsedContentType = new MediaType(contentType);
+            foreach (var mediaType in SupportedMediaTypes)
+            {
+                var parsedMediaType = new MediaType(mediaType);
+                if (parsedMediaType.HasWildcard)
+                {
+                    // For supported media types that are wildcard patterns, confirm that the requested
+                    // media type satisfies the wildcard pattern (e.g., if "text/entity+json;v=2" requested
+                    // and formatter supports "text/*+json").
+                    // Treat contentType like it came from a [Produces] attribute.
+                    if (parsedContentType.IsSubsetOf(parsedMediaType))
+                    {
+                        if (mediaTypes == null)
+                        {
+                            mediaTypes = new List<string>();
+                        }
 
-                // Confirm this formatter supports a more specific media type than requested e.g. OK if "text/*"
-                // requested and formatter supports "text/plain". Treat contentType like it came from an Accept header.
-                foreach (var mediaType in SupportedMediaTypes)
+                        mediaTypes.Add(contentType);
```

Comment:
The behaviour hasn't changed for this: previously it would happen if a developer explicitly added `"*/*"` to the `SupportedMediaTypes`, and with this PR, it still happens if (and only if) they do that.

However I recognise this is more of a hazard now we advertise support for wildcards on `SupportedMediaTypes` on output formatters. Fortunately, the only reasonable case is `*/*`, and not other patterns like `application/*`, because `contentType` only comes from `[Produces]` and similar, which explicitly prevent you from using wildcards, and from the new line 56 above, which adds `*/*`. I'll modify this to prevent `*/*`.